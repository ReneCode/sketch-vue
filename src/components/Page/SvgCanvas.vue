<template>
  <svg ref="svg" width="100%" height="100%">
    <g :transform="svgTransform">
      <!-- <image xlink:href="http://lorempixel.com/400/200/technics" x="100" y="80"/> -->
      <svg-item v-for="(item,index) in allItems" :key="index" :item="item" :iid="item.id" :class="item.selected? 'item-selected': 'item-normal'"></svg-item>
      <svg-item v-for="(item,index) in tmpItems" :key="index" :item="item" class="tmp" :iid="item.id"></svg-item>
    </g>
  </svg>
</template>

<script>
import Svg from "../../svg";
import SvgItem from "./SvgItem";

export default {
  name: "svg-canvas",
  props: ["allItems", "tmpItems"],
  components: {
    SvgItem
  },
  data() {
    return {
      svg: {}
    };
  },

  computed: {
    svgTransform() {
      if (this.svg) {
        return this.svg.getSvgTransformString();
      } else {
        return "";
      }
    }
  },

  created() {
    this.svg = new Svg();
  },

  mounted() {
    this.svg.init(this.$refs.svg);
  },

  beforeDestroy() {
    this.svg.exit();
  },

  methods: {}
};
</script>

<style>
svg {
  background-color: #f0f0f0;
}

.item-normal {
  stroke: #630;
  stroke-width: 2px;
  opacity: 0.9;
}

.item-selected {
  stroke: #222;
  stroke-width: 2px;
  stroke-dasharray: 5;
  /* cursor: pointer; */
  opacity: 0.8;
  animation: dash-rotate 0.5s linear infinite;
}

@keyframes dash-rotate {
  from {
    stroke-dashoffset: 10;
  }
  to {
    stroke-dashoffset: 0;
  }
}

.tmp {
  /* fill: #bbf; */
  stroke: #33c;
  cursor: pointer;
  opacity: 0.3;
}
</style>
