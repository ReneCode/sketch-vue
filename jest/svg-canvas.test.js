import { mount } from 'vue-test-utils'
import SvgCanvas from '../src/components/Page/SvgCanvas'
import ItemCircle from '../src/models/item-circle'

import Point from '../src/models/point'

describe('SvgCanvas', () => {
  let cmp

  beforeEach(() => {
    cmp = mount(SvgCanvas)
  })

  it('renders empty svgu', () => {
    expect(cmp.html()).toMatchSnapshot()
  })

  it('renders svg with allItems = circle', () => {
    const allItems = [];
    allItems.push(new ItemCircle(new Point(50, 20), 40));
    cmp.setProps({ allItems: allItems })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('renders svg with allItems = selected circle', () => {
    const allItems = [];
    const item = new ItemCircle(new Point(50, 20), 40);
    item.selected = true;
    allItems.push(item);
    cmp.setProps({ allItems: allItems })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('update renders when allItems changes', () => {
    const allItems = [];
    let item = new ItemCircle(new Point(50, 20), 40);
    allItems.push(item);
    cmp.setProps({ allItems: allItems })
    item.setRadius(30);
    cmp.update();
    expect(cmp.html()).toMatchSnapshot()
  })
})
