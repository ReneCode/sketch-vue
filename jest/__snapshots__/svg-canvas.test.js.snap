// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgCanvas renders empty svgu 1`] = `<svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"> </g></svg>`;

exports[`SvgCanvas renders svg with allItems = circle 1`] = `<svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"><circle cx="50" cy="20" r="40" iid="0" class="item-normal"></circle> </g></svg>`;

exports[`SvgCanvas renders svg with allItems = selected circle 1`] = `<svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"><circle cx="50" cy="20" r="40" iid="0" class="item-selected"></circle> </g></svg>`;

exports[`SvgCanvas update renders when allItems changes 1`] = `<svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"><circle cx="50" cy="20" r="30" iid="0" class="item-normal"></circle> </g></svg>`;
